tt_metal/tt_metal.cpp

```diff
@@ -171,6 +172,78 @@ std::vector<Device *> devices;
 
 }  // namespace device_pool
 
+namespace device_cpu_allocator {
+std::unordered_map<int, std::vector<uint32_t>> get_cpu_cores_per_numa_node(std::unordered_set<uint32_t> &free_cores) {
+    std::unordered_map<int, std::vector<uint32_t>> cpu_cores_per_numa_node = {};
+    if (numa_available() != -1) {
+        // Host has NUMA enabled. Group CPU IDs by the NUMA nodes they belong to.
+        for (int cpu = 0; cpu < numa_num_configured_cpus(); ++cpu) {
+            int node = numa_node_of_cpu(cpu);
+            if (cpu_cores_per_numa_node.find(node) == cpu_cores_per_numa_node.end()) {
+                cpu_cores_per_numa_node.insert({node, {}});
+            }
+            free_cores.insert(cpu);
+            cpu_cores_per_numa_node.at(node).push_back(cpu);
+        }
+    } else {
+        // Host does not have NUMA. Place all CPU Ids under a single node (0).
+        log_warning(tt::LogMetal, "Host does not use NUMA. May see reduced performance.");
+        for (int cpu = 0; cpu < sysconf(_SC_NPROCESSORS_ONLN); ++cpu) {
+            free_cores.insert(cpu);
+        }
+    }
+    return cpu_cores_per_numa_node;
+}
+
+int get_cpu_core_for_device_worker_thread(
+    int mmio_controlled_device_id,
+    const std::unordered_map<int, std::vector<uint32_t>> &cpu_cores_per_numa_node,
+    std::unordered_set<uint32_t> &free_cores) {
+    int core_assigned_to_device = 0;
+    if (numa_available() != -1) {
+        // Get NUMA node that the current device is mapped to through UMD
+        int numa_node_for_device = tt::Cluster::instance().get_numa_node_for_device(mmio_controlled_device_id);
+        if (cpu_cores_per_numa_node.find(numa_node_for_device) != cpu_cores_per_numa_node.end()) {
+            // NUMA node reported by UMD exists on host. Choose a core on this numa-node using round robin policy
+            int num_cores_in_numa_node = cpu_cores_per_numa_node.at(numa_node_for_device).size();
+            core_assigned_to_device =
+                cpu_cores_per_numa_node.at(numa_node_for_device).at(mmio_controlled_device_id % num_cores_in_numa_node);
+        } else {
+            // NUMA node reported by UMD does not exist on host. Use round-robin binding policy for this worker thread.
+            log_warning(
+                tt::LogMetal,
+                "NUMA node {} for device {} does not exist on host.",
+                numa_node_for_device,
+                mmio_controlled_device_id);
+            core_assigned_to_device = mmio_controlled_device_id % sysconf(_SC_NPROCESSORS_ONLN);
+        }
+    } else {
+        // System does not use NUMA. Use-round robin binding strategy.
+        core_assigned_to_device = mmio_controlled_device_id % sysconf(_SC_NPROCESSORS_ONLN);
+    }
+    free_cores.erase(core_assigned_to_device);
+    return core_assigned_to_device;
+}
+
+void bind_current_thread_to_free_cores(const std::unordered_set<uint32_t> &free_cores) {
```

Comment:
is this code called at init or during runtime?  I find it ironic that CPU_SET (etc) use a bitmask for cpus but we use an unordered set, these are orders of magnitude apart in terms of efficiency.  In general, I dislike how much we rely on sets and maps when bitmasks or arrays will work dramatically faster w/ dramatically less cache footprint.  I haven't wrapped my head around this enough to know if this is another instance but will raise the question here.

Reply:
This is during init only. We set thread affinities once at the start. I get your point though, we can use bit-masks here ... just makes some of the code look a bit ugly and wasn't initially intuitive so I used maps instead.